📇 Native Directory view

drewvolz opened the pull request:
Implementing a native directory client for iOS and Android using the directory API.

📸 Screenshots

~ | Android | iOS
--- | --- | ---
Initial | ![android initial](https://user-images.githubusercontent.com/5240843/190881462-3cd68458-6b82-4159-ada3-d96027450ae1.png) | ![ios initial](https://user-images.githubusercontent.com/5240843/190881463-ab81cf95-96fa-4021-952c-92fcde169589.png)
Results | ![android results](https://user-images.githubusercontent.com/5240843/190881465-fa57dfb9-c5e5-46e0-aacb-1190cb075947.png) | ![ios results](https://user-images.githubusercontent.com/5240843/190881467-3162e4f5-cdcb-4b08-a6b3-04d647694b1e.png)
Detail | ![android detail](https://user-images.githubusercontent.com/5240843/190881618-aa369cde-2af5-47e5-b771-dd68016aaed6.png) | ![ios detail](https://user-images.githubusercontent.com/5240843/190881470-2154bfb0-8cf3-47bd-83a9-4bbdb4cb535b.png)



<hr>

Known issues
- [x] flowtypes
- [x] error when searching about splitToArray being called on null
- [x] android searchbar eating the query while typing
- [x] android searchbar covering the first result
- [x] no idea what on-campus results will look like
- [x] no results not differentiated
- [x] unknown if error state presents itself
- [x] office view changes
  - [x] finish implementing the office section changes on iOS
  - [x] implement the office section changes to android
- [x] android searchbar is very broken on Orgs, Dictionary, and ClassLab Search, but works well here
- [x] department chips need to wrap when it gets really long
- [x] android searchbar is broken (fixed in https://github.com/StoDevX/AAO-React-Native/pull/3715)
- [x] office hours tweaks
- [ ] telling fetch() to not cache search results isn't working
- [ ] need to differentiate student vs fac/staff props

drewvolz committed:
WIP on student directory

drewvolz committed:
Update card and fetch imports

drewvolz committed:
Remove bugsnag and tracking

drewvolz committed:
Prevent cached results from returning

Co-authored-by: Hawken Rives <hawkrives@gmail.com>

drewvolz committed:
Remove matching on class year, building, and department

drewvolz committed:
Un-short-circuit navigation

drewvolz committed:
Use the 'u' flag due to require-unicode-regexp

drewvolz committed:
Attempt to setup the correct flow types

drewvolz committed:
Switch to functional setstate

drewvolz commented on source/views/directory/types.js:
```diff
@@ -0,0 +1,36 @@
+		city: string,
+		country: string,
+		state: string,
+		street: Array<string>,
```
not sure why this is an array on the backend?

hawkrives replied:
Because there are up to two street addresses and I didn't want to put two props in

drewvolz committed:
update type of homephone

hawkrives reviewed (commented):
Just remember that the API will be changing in as-yet-undetermined ways on either Monday or Tuesday

hawkrives commented on source/views/directory/list.js:
```diff
@@ -0,0 +1,244 @@
+			const sorted = sortBy(withSortableNames, '$sortableName')
+			const grouped = groupBy(sorted, '$groupableName')
+			this.setState(() => {
+				results: grouped
```
This is not returning an object; it's returning `undefined`, after evaluating a block with a `results:` label. You will need to wrap the {…} in parentheses in order for it to be parsed as an object.

drewvolz committed:
wrap setstate callbacks appropriately

drewvolz committed:
simplify contact locations (server handles most of that)

drewvolz committed:
update the API returned directory types

drewvolz committed:
WIP: partially update the iOS directory detail

rye reviewed (commented):
👀 All the eslint issues are just related to variables being created but never used. Remove them or prefix with `_`?

Then the `flow` stuff… I'm of no use on that.

hawkrives committed:
add react's eslint hooks plugin

hawkrives committed:
add dep on use-throttle

hawkrives committed:
add dep on react-async

hawkrives committed:
rework the directory view with hooks and flow-checkable platform-specific components

drewvolz committed:
nix use-throttle dependency

drewvolz committed:
switch to useDebounce module away from useThrottle

drewvolz committed:
check for generic errors and too short of search

hawkrives committed:
halve the debounce delay

it should still be fast enough for phones, without waiting for people to
push "search"

hawkrives committed:
show the searched query, instead of the typed one, when reporting no results

hawkrives committed:
rip out the android searchbar in favor of react-native-paper's searchbar

this breaks … oh, every other view with a search bar, but it's worth it

hawkrives committed:
redesign android directory list view

hawkrives committed:
redesign android directory detail view

hawkrives committed:
prettify

hawkrives committed:
fix wrapping departments on android

drewvolz committed:
add onPress and accessory to multi-line left detail cell

drewvolz committed:
office hours and on-leave helpers

drewvolz committed:
exported and expanded types for officeHours and onLeave

drewvolz committed:
show office hours and sabbatical info

drewvolz committed:
flow and prettier

hawkrives commented on modules/tableview/cells/multi-line-left-detail.js:
```diff
@@ -23,7 +25,13 @@ export class MultiLineLeftDetailCell extends React.PureComponent<LeftDetailProps
+			<Cell
+				accessory={accessory}
+				cellContentView={cellContent}
+				onPress={onPress ? onPress : null}
```
If you've defined `onPress` to be potentially null, why the runtime check here?

drewvolz replied:
To prevent the cell from being pressable

drewvolz replied:
Ah — I see your point now. I was thinking about a different file.

hawkrives commented on source/views/directory/detail-android.js:
```diff
@@ -93,11 +95,11 @@ export function DirectoryDetailView(props: Props) {
 				</View>
 
-				{officeHours ? (
+				{officeHours && (officeHours.prefix || officeHours.content) ? (
```
You should be able to just check if `officeHours` is null; the API will return `null` if the person doesn't have anything set.

drewvolz replied:
It’s more that I’m finding myself wanting to know if these individual props have content or not.

drewvolz committed:
remove nullity check for onpress

drewvolz committed:
remove additional checks for officehours props

drewvolz committed:
remove sabbatical checking

drewvolz committed:
abstract short room and description building into lib

drewvolz committed:
remove colon from prefix title if present

drewvolz committed:
handle html entities inside a contact title

drewvolz committed:
remove circular image radius from ios directory list

drewvolz committed:
unexport onleave type

rye commented on source/views/directory/lib.js:
```diff
@@ -0,0 +1,61 @@
+	departments: Array<Department>,
+	classYear: string,
+) => {
+	if (classYear !== null) {
```
The Rustacean in me is wary of a `null` comparison, especially one that's imbued with meaning as this one seems to be. Is there another way to do this check?

Also, what if `classYear` is non-`null` for fac/staff? Like if staff want to indicate that they were '69, we should support that.

(I think it's better to have @hawkrives figure out how to do this server-side first, personally.)

drewvolz committed:
Nix edithome (rebase artifact)

drewvolz committed:
Package updates from master

drewvolz committed:
Fixup merge conflict in package.json

drewvolz committed:
Updates to directory view and helpers

* Tsify all directory views
* Update directory to use proxied api call
* Add new return types from proxied call
* Update background color and text for empty search
* Wrap iOS search results with safearea
* Update tableview accessory types
* Update tableview onPress signature
* Format with prettier

drewvolz committed:
Nix directory helpers (moved to proxied endpoint)

drewvolz commented on source/views/types.ts:
```diff
@@ -1,11 +1,11 @@
+		getParam: (string, any) => any,
+		state: {params: P},
+	},
 }
```
I'm not sure how to make prettier happy with this file yet.

drewvolz committed:
Remove entity decoding for display title

drewvolz committed:
Move short location prop server-side

drewvolz committed:
Add newline back to lockfile

drewvolz committed:
Attempt at providing useful param names for navigation props

drewvolz committed:
More poking at fixing up view types

drewvolz committed:
Run prettier

rye commented on source/views/directory/detail.android.tsx:
```diff
@@ -0,0 +1,194 @@
+					? 'Call'
+					: `Call ${loc.buildingabbr} ${loc.room}`,
+			onPress: () => callPhone(loc.phone),
+		})),
```
Could this `map` body get extracted to a function?

rye commented on source/views/directory/detail.android.tsx:
```diff
@@ -0,0 +1,194 @@
+					},
+			  ]
+			: []),
+	]
```
In a similar vein to the previous idea on the `map` body — could this be a function as well?  Even if it is just an if-statement.

rye commented on source/views/directory/detail.android.tsx:
```diff
@@ -0,0 +1,194 @@
+		username,
+	} = props.navigation.state.params.contact
+
+	const [fabOpen, setFabOpen] = React.useState(false)
```
To make it more clear what `fabOpen` means, could you prefix it with `is`? (`isFabOpen`, and `setIsFabOpen`)

rye commented on source/views/directory/detail.android.tsx:
```diff
@@ -0,0 +1,194 @@
+	const showFab = fabActions.length >= 1
+
+	return (
+		<>
```
Was there any particular reason to not use `React.Fragment`?  (More just curious, but I had to look up this syntax.)

drewvolz replied:
No particular reason. I like the shorthand but if it is hard to read, I can switch it?

rye commented on source/views/directory/types.ts:
```diff
@@ -0,0 +1,70 @@
+	username: ?string
+	// added from ccc-server
+	description: string
+}
```
Typescript complains about these `?<type>` keys.

If these are optional (i.e. are optional if I'm constructing one of these), these could be written as e.g. `email?: string`.

If these are nullable, that should be explicit via `| null`, e.g. `email: string | null`.

drewvolz replied:
Oh yes, great suggestion. I forgot to modify these types!

rye commented on source/views/types.ts:
```diff
@@ -1,11 +1,14 @@
+		push: (routeName: string, params?: AnyObject) => mixed
+		goBack: () => mixed
+		getParam: (paramName: string, defaultValue: any) => any
+		state: {params: P}
```
TypeScript is complaining about these things too.  Is it possible to turn `mixed` into a union type at all?

rye commented on source/views/directory/detail.ios.tsx:
```diff
@@ -0,0 +1,157 @@
+								accessory={officeHours.href ? 'DisclosureIndicator' : undefined}
+								detail={officeHours.title}
+								onPress={
+									officeHours.href ? () => openUrl(officeHours.href) : null
```
TypeScript is complaining about this line and a couple of the other `onPress` lines.

```
No overload matches this call.
  Overload 1 of 2, '(props: LeftDetailProps | Readonly<LeftDetailProps>): MultiLineLeftDetailCell', gave the following error.
    Type '(() => Promise<any>) | null' is not assignable to type '(() => any) | undefined'.
      Type 'null' is not assignable to type '(() => any) | undefined'.
  Overload 2 of 2, '(props: LeftDetailProps, context: any): MultiLineLeftDetailCell', gave the following error.
    Type '(() => Promise<any>) | null' is not assignable to type '(() => any) | undefined'.ts(2769)
```

Do we need `null`? If so, the type definition needs to include that case. Or could the false case just be `undefined`?

(A side note: I'm not against giving anonymous functions a name and promoting them to functions — that helps make things traceable.)

drewvolz replied:
Good call. I'll opt to return `undefined` in the false case to match the updated type.

And I feel similarly about extracting inlined functions! I suppose I don't have a set rule other than thinking about how much work they do, how many lines they are, or if they can be shared.

drewvolz replied:
> and a couple of the other onPress lines.

As for other `onPress` return types like `<Cell />` (react-native-tableview-simple):

Return type is `() => void | false` but we tend to do other things like call `openUrl()` ... `() => Promise<any>`.

Other than forking the project and updating the type or patching locally, are there other options?

drewvolz replied:
What do we think of this solution? The type that ships for `CellInterface['onPress']` looks like an oversight as we know we are able to return, pass, and do things other than the provided signature `(() => any) | undefined)`.

I am suggesting a replacement `onPress` type because the one that ships with the library is too restrictive. I would prefer if we overwrite it and use our own less restrictive `onPress`.

```ts
type CellInterfaceModifiedType = Omit<CellInterface, 'onPress'> & {
	onPress?: () => void   // (or some variant of `void | undefined`)
}
```



Does this sound okay to you, @rye?

rye reviewed (dismissed):
Couple of ideas so far.  Looking good!

drewvolz committed:
Prefix fabOpen with 'is'

drewvolz committed:
Fix onPress return type for custom cell

drewvolz committed:
Extract phone and email FAB building actions

* Move logic into helpers
* Create new type to reflect what a FAB action returns
* Refactor email action for clarity

drewvolz committed:
Nix custom type and use exported type

drewvolz committed:
Clean up some of the directory types

drewvolz committed:
Check for url nullity before opening url

drewvolz committed:
Remove proxied directory searching in favor of on-device

We lose results that are provided on-campus if insert ourselves as the
endpoint, and that unfortunately means this endpoint is a little more
fragile while being potentially receiving a lot of traffic. The tradeoff
to show student results seems right, though.

This also moves the formatting of the data into helpers and out of the
proxied server.

drewvolz committed:
Merge branch 'master' into directory

drewvolz committed:
revert merge conflict which replaced the view type contents

drewvolz committed:
provide new navigation options for directory list and detail views

drewvolz committed:
update navigation routes, types, and view index for directory

drewvolz committed:
replace TopLevelViewPropsTypeWithParams with useRoute hook in detail

drewvolz committed:
add required visible prop to android FAB component

drewvolz committed:
add CellInterfaceModifiedType to allow a less restrictive onPress from
react-native-tableview-simple CellInterface

Replacing onPress type with a less restricted type of our own otherwise
we have to deal with issues around passing in functions and promises
that (i) cannot pass in parameters and (ii) must return void or false.

Otherwise we deal with type issues being flagged for valid callbacks.

drewvolz committed:
fix call while decoding entities in directory helpers

drewvolz committed:
pull out react-native-paper's FABGroupAction type so that we can use it

drewvolz committed:
refactor directory list

* useFetch hook to create useDirectory
* replace homegrown searchbar with native
* move "too short" and "empty search" errors
* add json error handling  (similar to menus)
* add new navigation routing
* fix prop types

drewvolz committed:
Merge branch 'master' into directory

drewvolz committed:
additional guards against blank location or phone data

hawkrives commented on source/navigation/types.tsx:
```diff
@@ -17,9 +17,14 @@ import type {
 } from '../views/menus/types'
+<<<<<<< HEAD
+import {DirectoryItem} from '../views/directory/types'
+||||||| fe7026171
```
Merge conflict

drewvolz replied:
Dang it, thanks.

hawkrives commented on modules/searchbar/searchbar-android.tsx:
```diff
@@ -0,0 +1,26 @@
+import * as React from 'react'
```
Didn't we get rid of this file?

drewvolz replied:
Somehow my merge really botched this. Git said it was removed when I committed this...

hawkrives commented on source/views/directory/types.ts:
```diff
@@ -0,0 +1,70 @@
+	thumbnail: string
+	title: string | null
+	username: string | null
+	// added from ccc-server
```
We're not proxying through ccc, are we?

drewvolz replied:
I had proposed a changeset to ccc to proxy before we realized we'd end up losing results that way. The proxied endpoint never was merged, but I can take a look to see if I forgot something on the client side. Good catch.

drewvolz replied:
We're good, looks like the comment remained but the logic is all on the client.

```tsx
//  https://github.com/StoDevX/AAO-React-Native/blob/directory/source/views/directory/helpers.ts#L80-L83

const shortRoomDetail = (loc: CampusLocation): string => {
	const shortRoom = `${loc.buildingabbr} ${loc.room}`.trim()
	return `${shortRoom ? `${shortRoom} • ` : ''}${loc.phone}`
}

// https://github.com/StoDevX/AAO-React-Native/blob/directory/source/views/directory/helpers.ts#L107-L110

const locations = campusLocations.map((loc: CampusLocation) => ({
	...loc,
	shortLocation: shortRoomDetail(loc),
}))
```

drewvolz committed:
fix merge conflicts from botched merge

- remove searchbar-anroid again
- clean up conflict markers in navigation types

drewvolz committed:
remove outdated comment because this is created on the client

drewvolz committed:
run prettier on ios directory detail

drewvolz committed:
Merge branch 'master' into directory

drewvolz committed:
update tableview import

drewvolz committed:
Merge branch 'master' into directory

drewvolz committed:
Merge branch 'master' into directory

drewvolz committed:
fix material icon names in android directory detail

drewvolz committed:
replace MaterialIcons for MaterialCommunity font
